### PR TITLE
bad use selector

### PR DESCRIPTION
**Pull request by chrisnojima**
- [x] transition chat message factories to not take the full `state`
- [x] add helper to pull out pieces chat factories need (`getMessageStateExtras`)
- [x] cleanup bad `useSelector`s

### CODE REVIEW

File: shared/actions/chat2/index.tsx [hunk 1]
```diff
@@ -1115,32 +1131,35 @@ function* loadMoreMessages(
 
   const loadingKey = Constants.waitingKeyThreadLoad(conversationIDKey)
   let calledClear = false
-  const onGotThread = (thread: string) => {
-    if (!thread) {
-      return
-    }
+  const onGotThread = (thread: string) =>
```
Comment (chrisnojima) [inline shared/actions/chat2/index.tsx:1134]: this fixes a kinda subtle bad behavior which is async callbacks plumbed through `incomingCallMap` that access a stale `state` from their parent. The fact that this was plumbed through `state` being passed to these message factories i think obfuscated it further

File: shared/actions/chat2/index.tsx [hunk 2]
```diff
@@ -1425,18 +1441,29 @@ const messageRetry = (action: Chat2Gen.MessageRetryPayload) => {
...
   const {conversationIDKey, viewType, fromMsgID} = action.payload
 
-  const onHit = (hit: RPCChatTypes.MessageTypes['chat.1.chatUi.chatLoadGalleryHit']['inParam']) => {
-    const message = Constants.uiMessageToMessage(state, conversationIDKey, hit.message)
-    return message
-      ? Saga.put(Chat2Gen.createAddAttachmentViewMessage({conversationIDKey, message, viewType}))
-      : []
-  }
+  const onHit = (hit: RPCChatTypes.MessageTypes['chat.1.chatUi.chatLoadGalleryHit']['inParam']) =>
+    Saga.callUntyped(function*() {
```
Comment (chrisnojima) [inline shared/actions/chat2/index.tsx:1451]: same problem here

File: shared/chat/audio/audio-recorder.native.tsx [hunk 3]
```diff
@@ -30,9 +30,7 @@ const AudioRecorder = (props: Props) => {
   const ampTracker = React.useRef(new AmpTracker()).current
   const [visible, setVisible] = React.useState(false)
   const [closingDown, setClosingDown] = React.useState(false)
-  const {audioRecording} = Container.useSelector(state => ({
```
Comment (chrisnojima) [inline shared/chat/audio/audio-recorder.native.tsx:33]: this makes a new object everytime, for no reason, so it thrashes on every store change

File: shared/chat/conversation/attachment-fullscreen/container.tsx [hunk 4]
```diff
@@ -17,11 +17,16 @@ type OwnProps = Container.RouteProps<{conversationIDKey: Types.ConversationIDKey
 const Connected = (props: OwnProps) => {
   const conversationIDKey = Container.getRouteProps(props, 'conversationIDKey', Constants.noConversationIDKey)
   const inOrdinal = Container.getRouteProps(props, 'ordinal', 0)
-  const state = Container.useSelector(s => s)
```
Comment (chrisnojima) [inline shared/chat/conversation/attachment-fullscreen/container.tsx:20]: basically the worst form of this. any state change period causes this to thrash

File: shared/chat/conversation/bot/install.tsx [hunk 5]
```diff
@@ -64,58 +64,46 @@ const InstallBotPopup = (props: Props) => {
...
-    settings,
-    teamID,
-    teamName,
-  } = Container.useSelector((state: Container.TypedState) => {
-    let inTeam: boolean | undefined
-    let isBot: boolean | undefined
-    let teamRole: TeamTypes.TeamRoleType | null | undefined
-    let teamName: string | null | undefined
-    let teamID: TeamTypes.TeamID = TeamTypes.noTeamID
-    let readOnly = false
```
Comment (chrisnojima) [inline shared/chat/conversation/bot/install.tsx:83]: this takes this mega useSelector and breaks it up so it doesn't entirely thrash. please look at this one carefully

File: shared/chat/conversation/bot/install.tsx [hunk 6]
```diff
@@ -64,58 +64,46 @@ const InstallBotPopup = (props: Props) => {
...
+
+  const meta = Container.useSelector(state =>
+    conversationIDKey ? state.chat2.metaMap.get(conversationIDKey) : undefined
+  )
+
+  // TODO will thrash every time
+  const commands = Container.useSelector(state => {
```
Comment (chrisnojima) [inline shared/chat/conversation/bot/install.tsx:73]: this one piece will thrash, someone should fix it

File: shared/constants/chat2/message.tsx [hunk 7]
```diff
@@ -19,6 +19,14 @@ import {assertNever} from '../../util/container'
 import invert from 'lodash/invert'
 import shallowEqual from 'shallowequal'
 
+export const getMessageStateExtras = (state: TypedState, conversationIDKey: Types.ConversationIDKey) => {
+  const getLastOrdinal = () =>
```
Comment (chrisnojima) [inline shared/constants/chat2/message.tsx:23]: this is a callback so we don't actually have to run it all the time. its only used for the pending messages

File: shared/chat/conversation/bot/install.tsx (hunk 5 above)
Comment (buoyad) [reply shared/chat/conversation/bot/install.tsx:83]: Looks like the types here could be cleaned up a bunch (not saying you should) but I couldn't find any logic differences here. Bot install seems to work fine :+1:

File: shared/teams/team/member/index.new.tsx [hunk 8]
```diff
@@ -67,6 +67,7 @@ const TeamMember = (props: OwnProps) => {
   React.useEffect(() => {
     dispatch(TeamsGen.createGetMemberSubteamDetails({teamID, username}))
   }, [teamID, username, dispatch])
+  // TODO this will keep thrasing
```
Comment (buoyad) [inline shared/teams/team/member/index.new.tsx:70]: This is all changing soon

**Pull request merged**